File: basic_dsa_two_hundred_twenty_seven.cpp
/**
 * Basic DSA - Find the Duplicate Number (LeetCode 287)
 * Find duplicate in array of n+1 integers in range [1,n] (Floyd cycle detection).
 */
#include <iostream>
#include <vector>

int findDuplicate(std::vector<int>& nums) {
    int slow = nums[0], fast = nums[0];
    do {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (slow != fast);
    slow = nums[0];
    while (slow != fast) {
        slow = nums[slow];
        fast = nums[fast];
    }
    return slow;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Find the Duplicate Number\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 3, 4, 2, 2};
    std::cout << "[1,3,4,2,2] -> " << findDuplicate(a1) << "\n";

    std::vector<int> a2 = {3, 1, 3, 4, 2};
    std::cout << "[3,1,3,4,2] -> " << findDuplicate(a2) << "\n";
}

File: basic_dsa_two_hundred_twenty_seven.py
"""
Basic DSA - 3Sum Closest
Given an integer array nums of length n and an integer target, find three integers
in nums such that the sum is closest to target. Return the sum.
Assume exactly one solution.
"""


def three_sum_closest(nums, target):
    """
    Sort + two pointers; track best sum by absolute difference.
    Time Complexity: O(n^2)
    Space Complexity: O(1) extra
    """
    nums.sort()
    best = nums[0] + nums[1] + nums[2]
    n = len(nums)
    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        l, r = i + 1, n - 1
        while l < r:
            s = nums[i] + nums[l] + nums[r]
            if abs(s - target) < abs(best - target):
                best = s
            if s == target:
                return s
            if s < target:
                l += 1
            else:
                r -= 1
    return best


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: 3Sum Closest")
    print("=" * 60)

    print(three_sum_closest([-1, 2, 1, -4], 1))  # 2
    print(three_sum_closest([0, 0, 0], 1))  # 0
